### PR TITLE
feat(dialog): support open with TemplateRef

### PR DESCRIPTION
**Pull request by mmalerba**

### CODE REVIEW

File: src/lib/dialog/dialog-container.ts [hunk 1]
```diff
@@ -61,21 +61,36 @@ export class MdDialogContainer extends BasePortalHost implements OnDestroy {
...
+    }
 
+    let attachedResult = this._portalHost.attachTemplatePortal(portal);
+    this._trapFocus();
+    return attachedResult;
+  }
+
+  /**
+   * Moves the focus inside the focus trap.
+   * @private
```
Comment (andrewseguin) [inline src/lib/dialog/dialog-container.ts]: Can remove the `@private` part of the jsdoc since its in the function signature

**Comment by andrewseguin** [comment]
Can you add a test for using template. LGTM otherwise, add merge ready when you are good with it

**Comment by fxck** [comment]
Is this replacement for https://github.com/angular/material2/pull/2853 then? Should I make a PR with stuff discussed here https://github.com/angular/material2/pull/2853#discussion_r98515217 after this PR and [this](https://github.com/angular/material2/pull/2825) PR gets merged(it needs rebase again btw @crisbeto)? How does merging workflow in this repository work anyway? Do you merge PRs labelled _merge ready_ once per two weeks?

**Comment by mmalerba** [comment]
@fxck Ah my bad, I didn't realize there was already a PR in progress for this. Let me take a look at yours. I'm hoping to get this feature in soon so I can use it for datepicker. Removed merge-ready until I have a chance to compare

**Comment by fxck** [comment]
@mmalerba the main difference is that I created a new method `openFromTemplate`, since I didn't know you could "cheat" `TemplatePortal` by giving it `null` as `ViewContainerRef`. I also exposed `backdropClick` and `escapePress` events for the reasons explained in the comment I linked previously. I didn't really finish the PR since I wanted to base it on @crisbeto's PR with `enter/exit` which changes quite a few things in overlay/backdrop. All in all, I'm fine with your PR replacing mine, I just really want those backdrop/escape events exposed so I can do this https://github.com/angular/material2/pull/2855

**Comment by mmalerba** [comment]
@fxck Ok, lets merge this one and then do another PR to add the observables you want.

**Comment by Splaktar** [comment]
It would have been really nice to include an update to the docs for this feature.

**Pull request merged**